applications/DG-Max/Algorithms/DivDGMaxEigenvalue.cpp

```diff
@@ -49,24 +49,190 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+            const Base::FaceMatrix& faceMatrix = face->getFaceMatrix(
+                DivDGMaxDiscretization<DIM>::FACE_STIFFNESS_MATRIX_ID);
+            LinearAlgebra::MiddleSizeMatrix block1, block2;
+            block1 = faceMatrix.getElementMatrix(Base::Side::LEFT,
+                                                 Base::Side::RIGHT);
+            block2 = faceMatrix.getElementMatrix(Base::Side::RIGHT,
+                                                 Base::Side::LEFT);
+
+            return std::make_pair(block1, block2);
+        });
+        kphaseshifts_ = builder.build(indexing_);
+    }
+
+    void initSolver() {
+        PetscErrorCode err;
+        err = EPSCreate(PETSC_COMM_WORLD, &solver_);
+        CHKERRABORT(PETSC_COMM_WORLD, err);
+        // Setting operators, but these will be set again after k-shifting
+        // Note the order Mx = (1/omega^2) Sx and NOT Sx = omega^2 M x and doing
+        // a spectral transformation.
+        // It may be possible to remove this relic with the current version of
+        // SLEPc.
+        err = EPSSetOperators(solver_, massMatrix_, stiffnessMatrix_);
+        CHKERRABORT(PETSC_COMM_WORLD, err);
+
+        err = EPSSetWhichEigenpairs(solver_, EPS_LARGEST_MAGNITUDE);
+        CHKERRABORT(PETSC_COMM_WORLD, err);
+        DGMaxLogger(INFO, "Eigenvalue solver configured");
+    }
+
+    Utilities::GlobalIndexing indexing_;
+    Utilities::GlobalPetscMatrix stiffnessMatrix_;
+    Utilities::GlobalPetscMatrix massMatrix_;
+    /// Vector used as intermediate storage and sample
+    Utilities::GlobalPetscVector tempVector_;
+
+    /// exp(ikx) shifts used in the stiffness matrix
+    DGMax::KPhaseShifts<DIM> kphaseshifts_;
+
+    LinearAlgebra::SmallVector<DIM> kpoint_;
+
+    /// Eigenvalue solver
+    EPS solver_;
+
+    /// Number of converged eigenpairs
+    PetscInt numberOfConvergedEigenpairs;
+    /// Number of vectors available in eigenvectors_, this may be larger than
+    /// the number of converged eigenvalues.
+    std::size_t numberOfEigenvectors_;
+    /// Storage for eigenvectors
+    Vec* eigenvectors_;
+    /// Storage for the most recent eigenvalues
+    std::vector<PetscScalar> eigenvalues_;
+};
 
 template <std::size_t DIM>
-struct DivDGMaxResult : AbstractEigenvalueResult<DIM> {
+class DivDGMaxResult : public AbstractEigenvalueResult<DIM> {
+   public:
+    DivDGMaxResult(const Workspace<DIM>& workspace) : workspace_(workspace) {
+        frequencies_.resize(workspace.numberOfConvergedEigenpairs);
+        for (std::size_t i = 0; i < frequencies_.size(); ++i) {
+            frequencies_[i] =
+                1. / std::sqrt(PetscRealPart(workspace_.eigenvalues_[i]));
+        }
+    };
+
     std::vector<double> getFrequencies() final { return frequencies_; }
 
     const LinearAlgebra::SmallVector<DIM>& getKPoint() const final {
-        return kpoint_;
+        return workspace_.kpoint_;
     }
 
+   private:
+    const Workspace<DIM>& workspace_;
```

Comment:
It is created in the solver. Then the result of a solve is given a reference (by the solve method) to the workspace so that it can access the actual results.